Add steam to latest

bloop16 opened the pull request:
Please add my adapter ioBroker.steam to latest.

*This pull request was created by https://www.iobroker.dev 20f0116.*

Commit:
Added steam to sources-dist.json

mcm1957 commented:
reminder 27.4.2025

mcm1957 commented:
@bloop16 
Please note the above notice:

>Important:
>
>To verify the object structure of this adapter during REVIEW please export the object structure of a working installation and attach the file to this PR. You find a guide how to export the object struture here: https://github.com/ioBroker/ioBroker.repochecker/blob/master/OBJECTDUMP.md

bloop16 commented:
Hope thats correct:

[steam.0.json](https://github.com/user-attachments/files/19897333/steam.0.json)

mcm1957 commented:
@bloop16 

First of all - THANK YOU for the time and effort you spend to maintain this adapter.

I would like to give some feedback based on my personal oppinion. @Apollon77 might have additional suggestions or even a different oppinion to one or the other statement. Please feel free to contact him if you cannot follow my suggestions or want to discuss some special aspects.
  
- [ ] package.json must specifiy required nodejs version

  Please adapt engines clause at package.json. New aapters should request node 20 minimum.
  ```
    "engines": {
    "node": ">=20"
  },
  ```

- [ ] unsual function '_'

  Why do you use a function _to process parameters? I suggest to use standard js functionality. So i.e. instead of using
  "this.log.info(this._('Need to resolve Steam ID for %s', steamName));" simply write
  "this.log.info(`Need to resolve Steam ID for ${steamName}`);". (Note backticks). This makes the code much more easy to read.

- [ ] onStateChange handler

  You implemented an onStateChange handler. But I did not see any subscribe operation. Are you sure this is working?

- [ ] info object

  The (standard) info object is not created anymore. It seems to get lost from io-package.json. Please add in front of info.connection object. (see as an example https://github.com/iobroker-community-adapters/ioBroker.snmp/blob/f41fe7362e256edcdc6f8a90cc65b2226bf3ebf0/io-package.json#L285)

- [ ] reevaluate state roles

  Only the values specified here (https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/stateroles.md) may be used as state roles. Do not invent new names as this might disturb the functionalyity of other adapters or vis.

  In addition the roles MUST match the read/write functionality. As an example a role value.* requires that the state is a readable state. Input states (write only) should have roles like level.*. Please read the description carefully. States with role 'button' should (must) have attribute 'common.read=false' set. A button ( "Taster" in german only triggers when you press but else has no state), so it should also be not readable. This is also like HomeMatic does it. A switch has clear states true or false and so can be read.

  Please avoid using general roles (state, value) whnever a dedicated role exists.

  Some examples:
  - text.url would be a good rule for the http(s):// urls located at several places
  
- [ ] consider using adapter.setTimeout / this.setTimeout instead of (standard) setTimeout
  
  The adapter package provides wrapper routines for native setTimeout, setInterval, clearTimeout and clearInterval. Using those routines ensures that timers are cancelled on on load. Additional checks on thomse limits might be performed, too. So consider replacing native setTimeout/clearTimeout by adapter.setTimeout/adapter.clearTimeout or this.setTimeout/this.clearTimeout. The same refers to setInterval/clearInterval.

  Note: This will NOT work inside onUnload at timers are blocked at during exit processing.

- [ ] adapt testing to supported node releases

  Node 18.x must be dropped together with adapting engines clause.
  
Thanks for reading and evaluating this suggestions.
McM1957

**Please add a comment when you have reviewed and fixed the suggestionsor at least commented the suggestions and you think the adapter is ready for a re-review!**


reminder 15.5.2025

bloop16 commented:
@mcm1957 
Thanks for assisting me with my first adapter and providing valuable insights!
Most of the code was written with AI, and adding and removing sections seemed to have introduced some unused elements. I have now thoroughly cleaned the code and reviewed your suggestions:

- Updated `package.json` to specify Node.js version 20 as minimum requirement.

- unsual function `_()`: Removed Custom `_()` Function.

- `onStateChange`: Implemented the `onStateChange` handler to track game state changes when gameID is modified (isPlaying). It appears to be working correctly.

- Adjusted the `info` object for improved consistency based on your recommendations.

- Re-evaluated and corrected state roles.

- Replaced standard `setTimeout` with `adapter.setTimeout` / `this.setTimeout` for improved lifecycle management.

- fully cleaned the code

I hope everything is functioning properly now! I'm still learning a lot about coding and related concepts, and I truly appreciate your support.

Thanks for helping!
Martin

bloop16 commented:
@mcm1957
Hi,
Is there still anything missing to get project to latest?

Thanks for your time!

mcm1957 commented:
nothing else then some time.
sorry

bloop16 commented:
Hi,
Is it correct that the `onStateChange` method in an adapter is automatically triggered by ioBroker whenever a state within my adapter's namespace gets changed? So I don't need an explicit `subscribeStates()` call for these?
I am using `onStateChange` to change state `isPlaying` on each game when `currentGame` and/or `currentGameAppId` have changed. It seams to work correct in the Adapter so everything should be fine.

[steam.0.json](https://github.com/user-attachments/files/20736483/steam.0.json)

Anyway could find a small issue with the `currentGamAppId`

mcm1957 commented:
@bloop16 

- [ ] State roles still contain invalid rolse. Please check your state roles against   https://www.iobroker.net/#en/documentation/dev/stateroles.md

  Examples (might not be a complete list!)
  - url.avatar is invalid 
    https://github.com/bloop16/ioBroker.steam/blob/582f626ae9b7d98c5f1b97444816a5b53a7c005f/main.js#L575  
    https://github.com/bloop16/ioBroker.steam/blob/582f626ae9b7d98c5f1b97444816a5b53a7c005f/io-package.json#L235  
  - value.number is invalid
    https://github.com/bloop16/ioBroker.steam/blob/582f626ae9b7d98c5f1b97444816a5b53a7c005f/main.js#L604  
    https://github.com/bloop16/ioBroker.steam/blob/582f626ae9b7d98c5f1b97444816a5b53a7c005f/main.js#L1245
    https://github.com/bloop16/ioBroker.steam/blob/582f626ae9b7d98c5f1b97444816a5b53a7c005f/main.js#L1288  
    https://github.com/bloop16/ioBroker.steam/blob/582f626ae9b7d98c5f1b97444816a5b53a7c005f/main.js#L1296
  - maybe more...  
  [This remark must be fixed]
  
- [ ] unneded getState
  At https://github.com/bloop16/ioBroker.steam/blob/582f626ae9b7d98c5f1b97444816a5b53a7c005f/main.js#L35 / ensureState you use getState and check if the state value has changes. This could be done with less overhead by using setStateChanged(Async). See https://github.com/ioBroker/ioBroker.js-controller/blob/6ade4045339082a9408e17e0871ad26dc41206ca/packages/adapter/src/lib/adapter/adapter.ts#L388. 
  [I do not take this remark as blocking, but please consider changing]

- [ ] onStateChange
  > Is it correct that the onStateChange method in an adapter is automatically triggered by ioBroker whenever a state within my adapter's namespace gets changed?
  no, onStateChange is only called for states explicitly subscribed (exakt id or matching wildcard). At least this the desired behaviour. So I'm quite sure your onStateChaneg code is never called. Did you check using debugger? I would suggest to ass a debug log to verify if and when onStateChecnge is called. And as wyou code seems to work without the onStateChange handler consider dactivating ist completly. 
  [I do not take this remark as blocking, but I suggest to analyse whats going on]


**Please ensure that you delete all Objects (deinstall the adapter at your testsystem) after fixing the roles so that the adapter creates new states with correct roles. Afte rthis please export objectview once more and post it here.**

reminder 24.6.2025

mcm1957 commented:
Uupps ...
How did you manage this:

❗ [E253] Version of [package.json](https://github.com/bloop16/ioBroker.steam/blob/master/package.json) (0.5.2) lower than latest version on [NPM](https://www.npmjs.com/package/iobroker.steam) (0.5.3)

Did you publish from local storage insted of doing it using workflow?
Anyway - please sync Github

mcm1957 commented:
RE-CHECK!

mcm1957 commented:
This adapter has been released to latest repository and should be visible within 24h maximum.

Please create a thread at https://forum.iobroker.net/category/91/tester titled like "Test Adapter " to collect some user feedback and provide a link to this topic when requesting addition to stable repository later.

Note: If an other testing topic already exists, it' OK to continue using this topic too.

If you do not have write access to TESTER Section of forum, please contact HOMORAN at forum using PN/CHAT
